keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+    // +optional
+    CredentialHashes map[string]ImagePullCredentials `json:"credentialHashes"`
+
+    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
+    // secrets that were used to pull the image
+    // +optional
+    KubernetesSecretCoordinates map[KubernetesSecretCoordinates]ImagePullSecret `json:"kubernetesSecretCoordinates"`
+
+    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
+    // credentials
+    // +optional
+    NodeWideCredentials bool `json:"nodeWideCredentials"`
+
+    // Anonymous is a flag denoting that the image was pulled without authentication.
+    // It is mutually exclusive to all the other fields of this structure.
+    // +optional
+    Anonymous bool `json:"anonymous"`
+}
+
+// TODO: the below are currently empty, would it be easier to extend these in the
+// `ImagePullCredentials` indexes than just mapping to an empty struct?
+type ImagePullCredentials struct {}
+type ImagePullSecret struct {}
+
+type KubernetesSecretCoordinates struct {
+    UID string       `json:"uid"`
+    Namespace string `json:"namespace"`
+    Name string      `json:"name"`
+}
+```
 
+**Examples:**<br>
+Contents of `/var/lib/kubelet/image_manager/pulling/sha256~9f023ac6b143be2e542ca832efa4f162392e3f88c6e9e77b149398d19e2ad1e2`:
+
+```json
+{
+  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
+  "kind": "ImagePullIntent",
+  "imageSpec": "docker.io/hello-world:latest",
+  "credentials": {
+    "anonymous": true
+  }
+}
 ```
-> {
->   "images": {
->     "sha256:eb6cbbefef909d52f4b2b29f8972bbb6d86fc9dba6528e65aad4f119ce469f7a": {
->       "authHash": { ** per review comment use SHA256 here vs hash **
->         "115b8808c3e7f073": {
->           "ensured": true,
->           "dueDate": "2023-05-30T05:26:53.76740982+08:00"
->         }
->       },
->       "name": "daocloud.io/daocloud/dce-registry-tool:3.0.8"
->     }
->   }
-> }
+
+Contents of `/var/lib/kubelet/image_manager/pulled/sha256~d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a`:
+
+```json
+{
+  "apiVersion": "imagemanager.kubelet.config.k8s.io/v1alpha1",
+  "kind": "ImagePulledRecord",
+  "imageId": "sha256:d2c94e258dcb3c5ac2798d32e1249e42ef01cba4841c2234249495f87264ac5a",
+  "credentials": {
+    "anonymous": true
+  }
+}
 ```
 
-See PR linked above for detailed design / behavior documentation.
+#### Kubelet Cache Housekeeping
+
+Upon kubelet start, the `pulling` part of the cache should be checked for dangling items.
+These would be the records of images being pulled that do not appear among images
+listed by container runtime at the start of the kubelet. If such a record exists, it
+should be pruned.
+
+**TODO**: how do we prevent the following race condition:
+1. kubelet prepares for a pull by writing an intent record
+2. kubelet issues an image pull request to the CRI
+3. kubelet restarts
+4. kubelet does not see the image so it removes the record from 1.
```

Comment:
I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.

Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

Reply:
Ok, worth following up but doesn't seem like we can rely on the annotations for now. The [other discussion](https://github.com/kubernetes/enhancements/pull/4789#discussion_r1754951743) about how to handle dangling intent-to-pull records simplifies and resolves the race between a background pull and a kubelet crash.

Reply:
> sometimes user may trigger a manual image pull using crictl to pull manually or to see the progress.

out of band pulls the kubelet never attempted to pull would be treated like preloaded images, and are out of scope for this enhancement

Reply:
> I was experimenting a little, but while I'm seeing the annotations being sent by the kubelet alongside the pull request, it doesn't seem that these would later be persisted - `crictl inspecti <imageref>` shows the imagespec, but the annotations are missing.
> 
> Perhaps @mikebrow might know if there's a way to do this, and if it would reliably work across different container runtimes.

They should be stored in meta, if not it is an error.